### PR TITLE
feat: rewrite and extension of features for virtual media

### PR DESCRIPTION
**Pull request by k3nd1k**
Mounting, unmounting ISOs, and booting to virtual media (Virtual CD). Works on both Dell and Supermicro.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -425,13 +427,23 @@ If you would like to check for any active virtual media you can run ```badfish``
 badfish -H mgmt-your-server.example.com -u root -p yourpass --check-virtual-media
 ```
 
+### Mount Virtual Media
+If you would like to mount an ISO from network you can run ```badfish``` with the ```--mount-virtual-media``` option which post a request for mounting the ISO virtual media (Virtual CD). Needs two arguments: the address and the path to the ISO.
+```bash
+badfish -H mgmt-your-server.example.com -u root -p yourpass --mount-virtual-media http://storage.example.com /folder/linux.iso
```
Comment (grafuls) [inline README.md:433]: small typo on the URL

File: src/badfish/main.py [hunk 2]
```diff
@@ -1353,115 +1223,174 @@ async def toggle_boot_device(self, device):
...
-                )
-
+                raise BadfishException("There was something wrong getting values for VirtualMedia")
+        return inserted
+
+    async def mount_virtual_media(self, host, path):
+        vm_config = await self.get_virtual_media_config()
+        _headers = {"Content-Type": "application/json"}
+        if self.vendor == "Supermicro":
+            _payload = {"Host": host, "Path": path, "Username": "", "Password": ""}
```
Comment (grafuls) [inline src/badfish/main.py:1292]: do you need to pass empty username and password here?

File: src/badfish/main.py [hunk 3]
```diff
@@ -1353,115 +1223,174 @@ async def toggle_boot_device(self, device):
...
+            _uri = "%s%s/Actions/IsoConfig.UnMount" % (
+                self.host_uri,
+                vm_config["config"],
+            )
+            _response = await self.post_request(_uri, payload="{}", headers=_headers)
+            if _response.status in [200, 202]:
+                self.logger.info("Image unmount operation was successful.")
+            else:
+                raise BadfishException("There was something wrong trying to unmount virtual media.")
+            _payload = {"Host": "", "Path": "", "Username": "", "Password": ""}
```
Comment (grafuls) [inline src/badfish/main.py:1333]: is this supposed to be totally empty?

**Review by grafuls — CHANGES_REQUESTED** [review]

File: src/badfish/main.py (hunk 3 above)
Comment (k3nd1k) [reply src/badfish/main.py:1333]: Yes, this is supposed to remove any information left from the previously mounted ISO. It doesn't automatically remove the information from the Supermicro's config for virtual media mounting on unmount request.

File: src/badfish/main.py (hunk 2 above)
Comment (k3nd1k) [reply src/badfish/main.py:1292]: Probably yes, it's mostly to ensure that any info isn't left from previous mounting operations.
And in [this](https://www.supermicro.com/manuals/other/RedfishRefGuide.pdf) (page 43) reference Redfish guide from Supermicro they themselves pass empty values on request for mounting ISO.

File: README.md (hunk 1 above)
Comment (k3nd1k) [reply README.md]: Changed with the latest [commit](https://github.com/redhat-performance/badfish/pull/324/commits/8f12e5b12bf23115b7d6e798205d0cb524354180).

File: src/badfish/main.py [hunk 4]
```diff
@@ -1353,115 +1223,175 @@ async def toggle_boot_device(self, device):
...
-
+                raise BadfishException("There was something wrong getting values for VirtualMedia")
+        return inserted
+
+    async def mount_virtual_media(self, path):
+        vm_config = await self.get_virtual_media_config()
+        _headers = {"Content-Type": "application/json"}
+        if self.vendor == "Supermicro":
+            split_path = str(path).split("/", 3)
+            _payload = {"Host": "/".join(split_path[:3]), "Path": "/" + split_path[-1], "Username": "", "Password": ""}
```
Comment (grafuls) [inline src/badfish/main.py:1293]: Afraid this code might be a little error prone with some corner case URLs. Safer to use `urllib.parse.urlparse`

``` bash
>>> from urllib.parse import urlparse
>>> urlparse(path)
ParseResult(scheme='http', netloc='storage.example.com', path='/folder/linux.iso', params='', query='', fragment='')
>>> urlparse(path).netloc
'storage.example.com'
>>> urlparse(path).path
'/folder/linux.iso'
```

**Review by grafuls — CHANGES_REQUESTED** [review]

File: src/badfish/main.py (hunk 4 above)
Comment (k3nd1k) [reply src/badfish/main.py]: Fixed with [this commit](https://github.com/redhat-performance/badfish/pull/324/commits/e593766fc05c12e635ac5dd70c306df7e3c84a22).

**Review by grafuls — APPROVED** [review]

**Pull request merged**